nncf/sparsity/magnitude/algo.py

```diff
@@ -59,17 +57,14 @@ def set_sparsity_level(self, sparsity_level):
         if sparsity_level >= 1 or sparsity_level < 0:
             raise AttributeError(
                 'Sparsity level should be within interval [0,1), actual value to set is: {}'.format(sparsity_level))
-        self.sparsity_level = sparsity_level
+        self._set_masks_for_threshold(self._select_threshold(sparsity_level))
 
-        self.threshold = self._select_threshold()
-        self._set_masks_for_threshold(self.threshold)
-
-    def _select_threshold(self):
+    def _select_threshold(self, sparsity_level):
         all_weights = self._collect_all_weights()
         if not all_weights:
             return 0.0
         all_weights_tensor, _ = torch.cat(all_weights).sort()
-        threshold = all_weights_tensor[int(all_weights_tensor.size(0) * self.sparsity_level)].item()
+        threshold = all_weights_tensor[int((all_weights_tensor.size(0) - 1) * sparsity_level)].item()
```

Comment:
doesn't it contradict to the previous logic with sparsity rate?

Reply:
Wherein lies the contradiction? This function does more or less the same as before, with added benefit that now it won't fail with an out-of-bounds error for sparsity rate = 1.0

Reply:
technically, it won't sparsify all values, but will keep the maximum values. but sparsity rate=1 is a corner case which shouldn't happen in a real life, so maybe it's ok